### PR TITLE
Scrape section finals schedule

### PR DESCRIPTION
**Pull request by adityavidyadharan**
Necessary to resolve gt-scheduler/website#92.
Paired with PR in website repo: gt-scheduler/website#134

Parses the [Final Exam Matrix PDF](https://registrar.gatech.edu/files/202208%20Final%20Exam%20Matrix.pdf) distributed for each term by the Registrar's Office.

The resulting information is then used to go back through the originally scraped data and add the new finals information. Classes with a common test period are correctly updated

PDF addresses are taken from [`matrix.json`](https://github.com/adityavidyadharan/crawler/blob/b5493df7677eb0993421dab109d1b4e3d2432c81/src/matrix.json)

**New section format:**
<pre>
"A": [
    "80684",
    [
        [
            0,
            "MW",
            "Scheller College of Business 224",
            0,
            [
                "Nicole Lena Mackenzie (P)"
            ],
            0,
            <b>1, // Index to caches.finalDates
            0  // Index to caches.finalTimes</b>
        ]
    ],
    3,
    0,
    0,
    [
        0
    ],
    0
],
</pre>
**Update to caches**
```
"caches": {
    "finalTimes": [
        "11:20 am - 2:10 pm",
        "2:40 pm - 5:30 pm",
        "6:00 pm - 8:50 pm",
        "8:00 am - 10:50 am"
    ],
    "finalDates": [
        "Dec 08, 2022",
        "Dec 09, 2022",
        "Dec 12, 2022",
        "Dec 13, 2022",
        "Dec 14, 2022",
        "Dec 15, 2022"
    ]
}
```
## Sample Files
 - Complete JSON for the 2022 Fall Semester in the updated format can be found [here](https://adityavidyadharan.github.io/crawler/202208.json)
 - CSV of the parsed Finals PDF can be found [here](https://adityavidyadharan.github.io/crawler/202208_Finals.csv)
 
 ## Parser Output
 
### General Schedule Parsing
**Example:**
|                          index| finalDate    | finalTime          |
|:------------------------------|:-------------|:-------------------|
| ('TR', '8:00 am - 9:15 am')   | Dec 15, 2022 | 8:00 am - 10:50 am |
| ('T', '8:00 am - 10:45 am')   | Dec 15, 2022 | 8:00 am - 10:50 am |
| ('TWRF', '8:25 am - 9:15 am') | Dec 15, 2022 | 8:00 am - 10:50 am |
| ('TRF', '8:25 am - 9:15 am')  | Dec 15, 2022 | 8:00 am - 10:50 am |
| ('TR', '8:25 am - 9:15 am')   | Dec 15, 2022 | 8:00 am - 10:50 am |

### Common Schedule Parsing

**Example:**

| Course    | Date         | Time              |
|:----------|:-------------|:------------------|
| ECE 3710  | Dec 08, 2022 | 6:00 pm - 8:50 pm |
| MATH 1551 | Dec 08, 2022 | 6:00 pm - 8:50 pm |
| MATH 1552 | Dec 08, 2022 | 6:00 pm - 8:50 pm |
| PHYS 2211 | Dec 09, 2022 | 6:00 pm - 8:50 pm |
| PHYS 2212 | Dec 12, 2022 | 6:00 pm - 8:50 pm |
| MATH 1553 | Dec 13, 2022 | 6:00 pm - 8:50 pm |
| MATH 1554 | Dec 13, 2022 | 6:00 pm - 8:50 pm |

### CODE REVIEW

**Review by jazeved0 — CHANGES_REQUESTED** [review]
## General comments

Thanks so much for contributing to this project! This is a really neat implementation to handle the problem of final exam matrix information not being available in a machine-readable structured way.

I also really like that the final exam matrix data is available in its own artifact (i.e. at [https://adityavidyadharan.github.io/crawler/202208_Finals.csv](https://adityavidyadharan.github.io/crawler/202208_Finals.csv)). This hopefully can be useful to other students wishing for a structured version of the final exam matrix data.

## Changes requested

I added a couple of inline comments, and I also have a few other sets of minor changes I'd like to see. For these other changes, let me preface these by saying: let me know if you disagree with any of the changes, and also if you need help/want to pair on any of them (especially, the TypeScript code, since I'm familiar with it).

### Updating the crawler code and schema

With the way that the code is currently written, the main TypeScript crawler CLI now produces an "un-augmented" JSON output for each term that then gets augmented with final exam data in `Revise.py` (let me know if my understanding is incorrect). The overall approach makes sense, but I think there is room for improvement in that the schema of the "un-augmented" JSON output and the "augmented" JSON output is different (there are new fields in the caches, and each section tuple has two more fields at the end). I think that this PR should also include changes to the base schema of the output JSON and the code that generates it in the main TypeScript crawler so that the schema is the same before and after augmenting using `Revise.py` (instead, the script could just add the data in a way that *enhances* the data-set). This would be useful for local development purposes (i.e. not needing to run the script/have Python installed and still get valid crawler data that can be used with the frontend). It would also make the crawler process more resilient if the Revise script fails for some reason and doesn't attach the data (but also doesn't fail the CI/CD job).

My outline for the a simple set of steps to do this is:
1. Update the `types.ts` file to include the relevant schema updates to the [`Caches` type](https://github.com/gt-scheduler/crawler/blob/main/src/types.ts#L30) and the [`Section` type](https://github.com/gt-scheduler/crawler/blob/main/src/types.ts#L109). When you do this, please also include comments describing the data (similar to what is in that file for the other fields)-- basically include what you already have in the PR description, and with a little more detail as needed (important: make sure to document that `-1` is the invalid/null value for the cache indices on `Section`).
2. Bump the [`CURRENT_VERSION` constant](https://github.com/gt-scheduler/crawler/blob/main/src/index.ts#L27) to reflect new schema changes. This can be used to ensure that the frontend can still work with old versions of the crawled data (since the crawler has data for [older terms](https://gt-scheduler.github.io/crawler/index.json)) by switching on the `version` field.
3. Update the generation of the the crawler data to include empty caches for the final matrix-related data ([link to relevant code](https://github.com/gt-scheduler/crawler/blob/main/src/steps/parse.ts#L59)) and invalid/`-1` indices on all Sections ([link to relevant code](https://github.com/gt-scheduler/crawler/blob/main/src/steps/parse.ts#L172)). This should only compile (using `yarn typecheck`) once the code has been updated to add these empty fields (assuming the `types` were updated in the earlier step).
4. Modify the `Revise.py` script to update the indices on the section objects instead of appending them. As a result, the script shouldn't materially change the schema of the JSON data at all, and the data should be able to be used on the frontend both before and after the Revise script runs.

### Code documentation

Could you please add some additional code documentation on a few different areas related to this PR:
1. In the `Revise` script, if you could add some documentation (either using comments or docstrings (`"""..."""`)) on:
    - What files the script expects to exist, and where (i.e. `./data/YYYYMM.json`), and what they contain
    - What files the script produces, and where (i.e. `./data/YYYYMM.json` but updated, `./data/YYYYMM_Finals.csv`), and what they contain (especially for the CSV file, some breif description of the schema would be helpful)
    - A brief description of what the script does and how/where it fits into the overall crawling process .
2. In the `Parse` script, if you could add some documentation (either using comments or docstrings (`"""..."""`)) on:
    - A brief description of what the main class does (i.e. "uses tabula to parse the final exam matrix PDF into a pandas DataFrame with columns/indices XYZ..." or similar)
    - Any short comments that you think might be useful for understanding the algorithm used. There are already some in the file, but perhaps some comments on what each function does would be useful. It's up to you what level of detail you think would be most useful, but I would say to use the criteria of "if I was unfamiliar with this code and had to fix a bug in a specific part of the algorithm in 2 years time, what would be most useful to me in order to identify the area of code to look at?". Feel free to use your best judgement.

### Local development/maintenence documentation

Could you please add some documentation in the README revolving around:
1. Steps for running the script locally (perhaps in a new section under "🚀 Running Locally") and how to observe the output (as in, a very breif summary of what the script changes).
2. Steps for updating the `matrix.json` file when a new term's data is released (perhaps in its own top-level section):
    - Where you obtain a new link to the PDF file (how did you originally find those links?)
    - When do the final exam matrices usually get released?
    - What this file does and why it needs to be updated.
    - What to look for in the CI/CD output to confirm that the final exam matrix was parsed successfully.

For context on the second step, it's on my TODO-list to write a more comprehensive document on the steps needed to maintain this project, so the information I'm asking for here would be really useful in the future once I work on writing that doc.

**Comment by adityavidyadharan** [comment]
Added a new wiki page which can be found [here](https://github.com/gt-scheduler/crawler/wiki/Finals-Scraping)

File: Finals.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
+# Finals Schedule Parser
```
Comment (jazeved0) [inline Finals.md:1]: I think if the docs are in the Wiki, they don't need to be checked in to source control as well. Moreover, having a single canonical version seems preferrable over having two separate copies of the same docs.

Let's remove this file from the PR (assuming the wiki page has the same contents)

File: README.md [hunk 2]
```diff
@@ -69,6 +69,39 @@ cat $(find . -type f -name "*.log" | sort -n | tail -1) | jq -cs '.'
 
 For some useful queries on the log data, see [📚 Useful queries on crawler logs](https://github.com/gt-scheduler/crawler/wiki/%F0%9F%93%9A-Useful-queries-on-crawler-logs).
 
+### Using the Python Finals Data Scraper
+
+Install the necessary Python modules with the included `requirements.txt` file
```
Comment (jazeved0) [inline README.md:74]: At the risk of being overly verbose, I'd like to also include the required min version of Python. (I based this on the GitHub Actions config, let me know if this is wrong)

```suggestion
First, ensure [Python 3.9 or newer](https://www.python.org/downloads/) is installed. Then, install the necessary Python modules with the included `requirements.txt` file:
```

File: README.md [hunk 3]
```diff
@@ -69,6 +69,39 @@ cat $(find . -type f -name "*.log" | sort -n | tail -1) | jq -cs '.'
...
+The Registrar publishes a PDF with the Finals schedule at the start of each semester.
+The page with the PDF for the Fall 2022 semester can be found [here](https://registrar.gatech.edu/info/final-exam-matrix-fall-2022)
+
+The `matrix.json` file contains a mapping from term to the pdf file.
+<br>The key is one of the terms identified by the scraper [here](https://gt-scheduler.github.io/crawler/202008.json).
+<br>The value is the direct address for the PDF file such as [this](https://registrar.gatech.edu/files/202208%20Final%20Exam%20Matrix.pdf)
+
+This mapping needs to be updated each semester when a new schedule is posted
+
+# TODO:
```
Comment (jazeved0) [inline README.md:101]: I think this might be a typo/temporary line?

File: src/types.ts [hunk 4]
```diff
@@ -62,6 +62,14 @@ export interface Caches {
...
    */
   locations: Location[];
+  /**
+   * List of the all the dates on which finals are happening
+   */
+  finalDates: Date[];
+  /**
+   * List of the time blocks for finals
+   */
+  finalTimes: string[];
```
Comment (jazeved0) [inline src/types.ts:72]: Grabbing latest docs from https://github.com/gt-scheduler/website/pull/134 to keep them in sync.

```suggestion
  /**
   * List of the all the dates on which finals are happening
   * Example date: Aug 02, 2022
   */
  finalDates: Date[];
  /**
   * List of the time blocks for finals
   * Example time: 11:20 am - 2:10 pm
   */
  finalTimes: string[];
```

File: src/types.ts [hunk 5]
```diff
@@ -176,7 +184,19 @@ export type Meeting = [
...
+   * specifying the date at which the final is
+   * -1 when no match could be found
+   */
+  finalDateIndex: number,
+  /**
+   * integer index into caches.finalTimes,
+   * specifying the time at which the final is
+   * -1 when no match could be found
+   */
+  finalTimeIdx: number
```
Comment (jazeved0) [inline src/types.ts:199]: Grabbing latest docs from https://github.com/gt-scheduler/website/pull/134 to keep them in sync.

```suggestion
  /**
   * integer index into caches.finalDates,
   * specifying the date at which the final is
   * -1 when no match could be found and
   * as a default value
   */
  finalDateIndex: number,
  /**
   * integer index into caches.finalTimes,
   * specifying the time at which the final is
   * -1 when no match could be found
   * and as a default value
   */
  finalTimeIdx: number
```

**Review by jazeved0 — CHANGES_REQUESTED** [review]
The wiki page looks great 🤩! Thanks for making all of the requested changes. I just had a few really minor final cleanups before I merge.

File: Finals.md (hunk 1 above)
Comment (adityavidyadharan) [reply Finals.md:1]: Accidentally committed this file, it was just meant to be where I could write a draft for the wiki

**Review by jazeved0 — APPROVED** [review]

**Pull request merged**